core/src/main/java/org/apache/calcite/rex/RexSimplify.java

```diff
@@ -674,135 +675,318 @@ private RexNode simplifyCoalesce(RexCall call) {
   }
 
   private RexNode simplifyCase(RexCall call, RexUnknownAs unknownAs) {
-    final List<RexNode> operands = call.getOperands();
-    final List<RexNode> newOperands = new ArrayList<>();
-    final Set<String> values = new HashSet<>();
-    for (int i = 0; i < operands.size(); i++) {
-      RexNode operand = operands.get(i);
-      if (RexUtil.isCasePredicate(call, i)) {
-        if (operand.isAlwaysTrue()) {
-          // Predicate is always TRUE. Make value the ELSE and quit.
-          newOperands.add(operands.get(++i));
-          if (unknownAs == FALSE && RexUtil.isNull(operands.get(i))) {
-            values.add(rexBuilder.makeLiteral(false).toString());
+    List<CaseBranch> inputBranches =
+        CaseBranch.fromCaseOperands(rexBuilder, new ArrayList(call.getOperands()));
+
+    // run simplification on all operands
+    RexSimplify branchSimplifier = this;
+    RelDataType branchType = call.getType();
+
+    List<CaseBranch> branches = new ArrayList<>();
+    for (CaseBranch branch : inputBranches) {
+      // simplify the condition
+      RexNode newCond = branchSimplifier.simplify(branch.cond, RexUnknownAs.FALSE);
+
+      // use the condition to simplify the branch
+      RexNode value = branch.value;
+      RexNode newValue = branchSimplifier.simplify(value, unknownAs);
+
+      branches.add(new CaseBranch(newCond, newValue));
+    }
+
+    // remove branches with invalid conditions
+    branches.removeIf(branch -> branch.cond.isAlwaysFalse());
+
+    // delete all branches after the first AlwaysTrue
+    for (int i = 0; i < branches.size(); i++) {
+      CaseBranch branch = branches.get(i);
+      if (branch.cond.isAlwaysTrue()) {
+        branches.subList(i + 1, branches.size()).clear();
+        break;
+      }
+    }
+
+    branches = compactBranchesWithTheSameConclusion(branches);
+
+    // collect cardinality of values
+    Set<String> values =
+        branches.stream().map(branch -> branch.value.toString()).collect(Collectors.toSet());
```

Comment:
Is it really required?
What is it useful for?

Reply:
this was present before; in a different form ; the goal of this is to collect all branch values; and if every branch came to the same conclusion - it removes the whole `case`.

I kept it since the `compactBranchesWithTheSameConclusion` may not compact some cases because of the need of safety - and this might still be relevant.

Reply:
> and this might still be relevant.

This is not relevant.
You must not "optimize" non-safe case expressions.

Consider the same `case when x<>0 then true when 1/x>0 then true else true end`
It MUST throw an error, and you MUST NOT optimize the expression to `true`

Reply:
I understand your point; but removing this would be a feature removal - I think it will need it's own ticket - since the current change is a feature addition.

Reply:
It is not feature removal

Reply:
The code you add is just plain invalid.
What's the point of adding it?

Reply:
On top of that, the particular line won't optimize a single new case. What's the point of having dead code in the codebase?

Reply:
It was there before:
https://github.com/apache/calcite/blob/fb7914ec61fc57c8348d9da144e2fbd608c00c61/core/src/main/java/org/apache/calcite/rex/RexSimplify.java#L707
look for the "values" array; I've refactored it to be easily readable - and not hide it in the woods...and just added a comment about what it's doing..

Because of the above I feel it more like a feature removal and as such would need a separate ticket. Do you still think I should remove it?

Reply:
Of course I do